lib/Conversion/SMTToZ3LLVM/LowerSMTToZ3LLVM.cpp

```diff
@@ -0,0 +1,925 @@
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LowerSMTToZ3LLVMPass
+    : public circt::impl::LowerSMTToZ3LLVMBase<LowerSMTToZ3LLVMPass> {
+  using Base::Base;
+  void runOnOperation() override;
+};
+} // namespace
+
+void circt::populateSMTToZ3LLVMTypeConverter(TypeConverter &converter) {
+  converter.addConversion([](smt::BoolType type) {
+    return LLVM::LLVMPointerType::get(type.getContext());
+  });
+  converter.addConversion([](smt::BitVectorType type) {
+    return LLVM::LLVMPointerType::get(type.getContext());
+  });
+  converter.addConversion([](smt::ArrayType type) {
+    return LLVM::LLVMPointerType::get(type.getContext());
+  });
+  converter.addConversion([](smt::IntType type) {
+    return LLVM::LLVMPointerType::get(type.getContext());
+  });
+  converter.addConversion([](smt::SMTFuncType type) {
+    return LLVM::LLVMPointerType::get(type.getContext());
+  });
+  converter.addConversion([](smt::SortType type) {
+    return LLVM::LLVMPointerType::get(type.getContext());
+  });
+}
+
+void circt::populateSMTToZ3LLVMConversionPatterns(
+    RewritePatternSet &patterns, TypeConverter &converter,
+    SMTGlobalsHandler &globals, const LowerSMTToZ3LLVMOptions &options) {
+#define ADD_VARIADIC_PATTERN(OP, APINAME, MIN_NUM_ARGS)                        \
+  patterns.add<VariadicSMTPattern<OP>>(/*NOLINT(bugprone-macro-parentheses)*/  \
+                                       converter, patterns.getContext(),       \
+                                       globals, options, APINAME,              \
+                                       MIN_NUM_ARGS);
+  // Lower `smt.distinct` operations which allows a variadic number of operands
+  // according to the `:pairwise` attribute. The Z3 API function supports a
+  // variadic number of operands as well, i.e., a direct lowering is possible:
+  // ```
+  // Z3_ast Z3_API Z3_mk_distinct(Z3_context c, unsigned num_args, Z3_ast const
+  // args[])
+  // ```
+  // The API function requires num_args > 1 which is guaranteed to be satisfied
+  // because `smt.distinct` is verified to have > 1 operands.
+  ADD_VARIADIC_PATTERN(DistinctOp, "Z3_mk_distinct", 2);
+
+  // Lower `smt.and` operations which allows a variadic number of operands
+  // according to the `:left-assoc` attribute. The Z3 API function supports a
+  // variadic number of operands as well, i.e., a direct lowering is possible:
+  // ```
+  // Z3_ast Z3_API Z3_mk_and(Z3_context c, unsigned num_args, Z3_ast const
+  // args[])
+  // ```
+  // The API function requires num_args > 0. This is not guaranteed by the
+  // `smt.and` operation and thus the pattern will not apply when no operand is
+  // present. The constant folder of the operation is assumed to fold this to
+  // a constant 'true' (neutral element of AND).
+  ADD_VARIADIC_PATTERN(AndOp, "Z3_mk_and", 1);
+
+#undef ADD_VARIADIC_PATTERN
```

Comment:
Not sure what the CIRCT/MLIR/LLVM policy is for using macros, but this could possibly be just a templated lambda, or, if there are only two uses of `ADD_VARIADIC_PATTERN`, it might be more readable to just write out the `patterns.add ...` calls.

Reply:
I would opt for just leaving the macros for now until we update to C++20 and can then replace them with templated lambdas for the following reasons:
* I couldn't find anything in the LLVM coding standards and programmers manual that forbids macros.
* All the visitors in CIRCT use macros in this exact same way. They are very local and self-contained, which shouldn't lead to debugging or readability issues.
* AFAIK templated lambdas are a C++20 feature, but CIRCT uses C++17. 
* As you have already seen in the follow-up PR, quite a few lowering patterns use those macro shortcuts. Inlining all of them would lead to decreased readability and make it more cumbersome to, i.e., add a constructor argument in the future.

Reply:
Yeah makes total sense to me! And you're right, this looks very much like our established uses in the visitors :sweat_smile: